pil-analyzer/src/type_inference.rs

```diff
@@ -624,22 +623,59 @@ impl<'a> TypeChecker<'a> {
             }
             Expression::BlockExpression(_, BlockExpression { statements, expr }) => {
                 let original_var_count = self.local_var_types.len();
+                let empty_tuple = Type::Tuple(TupleType { items: vec![] });
+                let expected_empty_tuple = ExpectedType {
+                    ty: empty_tuple.clone(),
+                    allow_array: false,
+                };
                 for statement in statements {
                     match statement {
                         StatementInsideBlock::LetStatement(LetStatementInsideBlock {
                             pattern,
                             value,
                         }) => {
                             let value_type = if let Some(value) = value {
-                                self.infer_type_of_expression(value)?
+                                let statement_type = self.infer_type_of_expression(value)?;
+                                match self.lambda_kind {
+                                    Some(FunctionKind::Constr) => {
+                                        if (statement_type == self.statement_type.ty)
+                                            | (statement_type == empty_tuple)
+                                        {
+                                            statement_type
+                                        } else {
+                                            return Err(format!(
+                                                "Invalid statement of type {statement_type} inside Constr function."
+                                            ));
+                                        }
+                                    }
+                                    _ => {
+                                        if statement_type == self.statement_type.ty {
+                                            return Err(format!(
+                                                "Invalid statement of type {statement_type} inside Pure/Query function."
+                                            ));
+                                        }
+                                        statement_type
+                                    }
+                                }
                             } else {
                                 Type::Expr
                             };
                             self.expect_type_of_pattern(&value_type, pattern)?;
                         }
-                        StatementInsideBlock::Expression(expr) => {
-                            self.expect_type_with_flexibility(self.statement_type, expr)?;
-                        }
+                        StatementInsideBlock::Expression(expr) => match self.lambda_kind {
+                            Some(FunctionKind::Constr) => {
+                                self.expect_type_with_flexibility(&self.statement_type, expr).or_else(|err| {
```

Comment:
I don't feel comfortable with recovering from errors inside the type inference. It might be that type unification did some changes that we then would have to undo. It's better to infer the type and if it is the empty tuple, accept that, otherwise call `expect...` for the other type.